File: src/codeswarm/acp/collaboration.py
"""Shared public context for CodeSwarm collaboration coordinators."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Protocol, Sequence


class AgentContextLike(Protocol):
    def get_info(self) -> object: ...


STOP_TOKEN = "[CODESWARM:STOP]"
DEFAULT_STOP_ACKNOWLEDGMENT = "👍"
MAX_RELAY_RESPONSE_CHARS = 12_000
MAX_RELAY_HISTORY_CHARS = 24_000
MAX_RELAY_JOURNAL_CHARS = 48_000
MAX_RELAY_EVENTS = 200


@dataclass(frozen=True)
class RelayEvent:
    """One public conversation update that other agents may not have seen."""

    speaker: str
    text: str


@dataclass
class CollaborationContext:
    """Bounded public journal and prompt state shared across modes."""

    agent_count: int
    shared_task: str | None = None
    public_events: list[RelayEvent] | None = None
    seen_event_count: list[int] | None = None
    history_truncated: list[bool] | None = None
    turn_instructions: str = ""

    def __post_init__(self) -> None:
        self.public_events = list(self.public_events or [])
        self.seen_event_count = list(self.seen_event_count or [0] * self.agent_count)
        self.history_truncated = list(
            self.history_truncated or [False] * self.agent_count
        )
        self.ensure_agent_count(self.agent_count)

    def ensure_agent_count(self, count: int) -> None:
        while len(self.seen_event_count) < count:
            self.seen_event_count.append(0)
            self.history_truncated.append(False)
        self.agent_count = max(self.agent_count, count)

    def set_turn_instructions(self, instructions: str) -> None:
        self.turn_instructions = instructions.strip()

    def add_agent(self) -> None:
        self.ensure_agent_count(self.agent_count + 1)

    @staticmethod
    def compact_response(response: str) -> str:
        if len(response) <= MAX_RELAY_RESPONSE_CHARS:
            return response
        head_size = MAX_RELAY_RESPONSE_CHARS // 2
        tail_size = MAX_RELAY_RESPONSE_CHARS - head_size
        return (
            response[:head_size]
            + "\n\n[CodeSwarm omitted the middle of this response to protect context.]\n\n"
            + response[-tail_size:]
        )

    def record_event(self, speaker: str, text: str, active: Sequence[bool]) -> int:
        """Append one bounded public event and return its journal index."""
        self._prune_public_events(active)
        self.public_events.append(
            RelayEvent(speaker, self.compact_response(text).strip())
        )
        return len(self.public_events) - 1

    def _prune_public_events(self, active: Sequence[bool]) -> None:
        active_indices = [index for index, value in enumerate(active) if value]
        consumed = min(
            (self.seen_event_count[index] for index in active_indices),
            default=0,
        )
        if consumed:
            del self.public_events[:consumed]
            self.seen_event_count = [
                max(0, count - consumed) for count in self.seen_event_count
            ]

        retained_chars = sum(
            len(event.speaker) + len(event.text) for event in self.public_events
        )
        while self.public_events and (
            len(self.public_events) >= MAX_RELAY_EVENTS
            or retained_chars >= MAX_RELAY_JOURNAL_CHARS
        ):
            removed = self.public_events.pop(0)
            retained_chars -= len(removed.speaker) + len(removed.text)
            for index, count in enumerate(self.seen_event_count):
                if count:
                    self.seen_event_count[index] = count - 1
                elif index < len(active) and active[index]:
                    self.history_truncated[index] = True

    def unseen_updates(self, agent_index: int, *, excluding: int | None) -> str:
        """Render public events not included in this agent's prior prompts."""
        start = self.seen_event_count[agent_index]
        parts = [
            f"{event.speaker}:\n{event.text}"
            for index, event in enumerate(self.public_events[start:], start)
            if index != excluding and event.text
        ]
        rendered = "\n\n".join(parts)
        if self.history_truncated[agent_index]:
            marker = "[CodeSwarm omitted older unseen updates to protect context.]"
            rendered = f"{marker}\n\n{rendered}" if rendered else marker
            self.history_truncated[agent_index] = False
        if len(rendered) <= MAX_RELAY_HISTORY_CHARS:
            return rendered

        marker = "[CodeSwarm omitted older unseen updates to protect context.]"
        remaining = MAX_RELAY_HISTORY_CHARS - len(marker) - 2
        newest: list[str] = []
        used = 0
        for part in reversed(parts):
            added = len(part) + (2 if newest else 0)
            if used + added > remaining:
                break
            newest.append(part)
            used += added
        newest.reverse()
        return f"{marker}\n\n" + "\n\n".join(newest)

    @staticmethod
    def name(agent: AgentContextLike) -> str:
        return str(agent.get_info())

    def build_turn_prompt(
        self,
        task: str,
        context: str,
        *,
        active_agents: Sequence[AgentContextLike],
        previous_agent: AgentContextLike | None,
        unseen_updates: str = "",
        can_stop: bool = False,
        stop_token: str = STOP_TOKEN,
    ) -> str:
        previous = (
            "This is the first turn."
            if previous_agent is None
            else f"Previous participant: {self.name(previous_agent)}."
        )
        work_instruction = (
            self.turn_instructions
            or "Inspect the shared workspace and make useful progress."
        )
        updates = (
            "Conversation updates since your previous turn:\n"
            f"{unseen_updates}\n\n"
            if unseen_updates
            else ""
        )
        if can_stop:
            stop_instruction = (
                "You are reviewing another participant's answer. If it needs no "
                "meaningful correction or addition, reply with an optional "
                "acknowledgment emoji followed by "
                f"{stop_token} as the final line. If you provide no emoji, "
                f"CodeSwarm displays {DEFAULT_STOP_ACKNOWLEDGMENT}. If you add "
                "substantive content, do not use the token; let the next "
                "participant review your contribution."
            )
        else:
            stop_instruction = (
                f"Do not use {stop_token} on this turn. Your response must be "
                "offered to another participant for review, even when you are "
                "highly confident it is correct."
            )
        return (
            f"You are one participant in a {len(active_agents)}-agent "
            "automated collaboration.\n\n"
            f"Shared task:\n{task}\n\n"
            f"{updates}"
            f"Turn context:\n{context}\n\n"
            f"{previous}\n"
            f"{work_instruction}\n\n"
            f"{stop_instruction}\n"
            "The token is internal: CodeSwarm hides it from the conversation."
        )

File: crates/codeswarm-core/src/collaboration.rs
//! Bounded public context shared between sequential relay participants.

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct PublicEvent {
    pub speaker: String,
    pub text: String,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct CollaborationContext {
    events: Vec<PublicEvent>,
    seen: Vec<usize>,
    truncated: Vec<bool>,
}

impl CollaborationContext {
    pub fn new(agent_count: usize) -> Self {
        Self {
            events: Vec::new(),
            seen: vec![0; agent_count],
            truncated: vec![false; agent_count],
        }
    }

    pub fn record(&mut self, speaker: impl Into<String>, text: impl Into<String>, active: &[bool]) {
        self.prune(active);
        self.events.push(PublicEvent {
            speaker: speaker.into(),
            text: compact(text.into()),
        });
    }

    pub fn mark_seen(&mut self, slot: usize) {
        if let Some(seen) = self.seen.get_mut(slot) {
            *seen = self.events.len();
        }
    }

    pub fn unseen(&mut self, slot: usize) -> String {
        let start = self.seen.get(slot).copied().unwrap_or(self.events.len());
        let mut updates = self.events[start..]
            .iter()
            .filter(|event| !event.text.is_empty())
            .map(|event| format!("{}:\n{}", event.speaker, event.text))
            .collect::<Vec<_>>();
        if self.truncated.get(slot).copied().unwrap_or(false) {
            updates.insert(
                0,
                "[CodeSwarm omitted older unseen updates to protect context.]".into(),
            );
            self.truncated[slot] = false;
        }
        limit(updates, 24_000)
    }

    fn prune(&mut self, active: &[bool]) {
        let consumed = active
            .iter()
            .enumerate()
            .filter_map(|(slot, enabled)| {
                enabled.then_some(self.seen.get(slot).copied().unwrap_or(0))
            })
            .min()
            .unwrap_or(0);
        if consumed > 0 {
            self.events.drain(..consumed);
            for seen in &mut self.seen {
                *seen = seen.saturating_sub(consumed);
            }
        }
        while self.events.len() >= 200
            || self
                .events
                .iter()
                .map(|event| event.text.len())
                .sum::<usize>()
                >= 48_000
        {
            self.events.remove(0);
            for (slot, seen) in self.seen.iter_mut().enumerate() {
                if *seen > 0 {
                    *seen -= 1;
                } else if active.get(slot).copied().unwrap_or(false) {
                    self.truncated[slot] = true;
                }
            }
        }
    }
}

fn compact(text: String) -> String {
    const LIMIT: usize = 12_000;
    if text.len() <= LIMIT {
        return text;
    }
    let head = LIMIT / 2;
    format!(
        "{}\n\n[CodeSwarm omitted the middle of this response to protect context.]\n\n{}",
        &text[..head],
        &text[text.len() - (LIMIT - head)..],
    )
}

fn limit(mut updates: Vec<String>, limit: usize) -> String {
    let rendered = updates.join("\n\n");
    if rendered.len() <= limit {
        return rendered;
    }
    let marker = "[CodeSwarm omitted older unseen updates to protect context.]";
    let mut selected = Vec::new();
    let mut used = 0;
    while let Some(update) = updates.pop() {
        let added = update.len() + 2;
        if used + added > limit - marker.len() - 2 {
            break;
        }
        used += added;
        selected.push(update);
    }
    selected.reverse();
    format!("{marker}\n\n{}", selected.join("\n\n"))
}

#[cfg(test)]
mod tests {
    use super::CollaborationContext;

    #[test]
    fn only_unseen_public_text_is_sent_to_each_agent() {
        let mut context = CollaborationContext::new(2);
        context.record("Human", "task", &[true, true]);
        context.mark_seen(0);
        assert_eq!(context.unseen(0), "");
        assert_eq!(context.unseen(1), "Human:\ntask");
    }

    #[test]
    fn long_history_is_bounded_without_losing_recent_updates() {
        let mut context = CollaborationContext::new(1);
        for index in 0..250 {
            context.record("Agent", format!("reply {index}"), &[true]);
        }
        let unseen = context.unseen(0);
        assert!(unseen.contains("reply 249"));
        assert!(unseen.len() <= 24_000);
    }
}
